app/controllers/api/v1/base_controller.rb

```diff
@@ -0,0 +1,87 @@
+# frozen_string_literal: true
+
+class Api::V1::BaseController < ActionController::API
+  include Pundit
+  include CustomErrors
+
+  DEFAULT_PER_PAGE = 5
+
+  rescue_from ActionController::ParameterMissing do
+    api_error(status: 400, errors: "invalid parameters")
+  end
+
+  rescue_from Pundit::NotAuthorizedError do
+    unauthorized!
+  end
+
+  rescue_from MissingAuthHeader do
+    api_error(status: 401, errors: "Authentication header is missing")
+  end
+
+  rescue_from ActiveRecord::RecordNotFound do
+    api_error(status: 404, errors: "resource not found!")
+  end
+
+  rescue_from ActiveRecord::RecordInvalid do
+    api_error(status: 422, errors: "resource invalid!")
+  end
+
+  rescue_from UnauthenticatedError do
+    unauthenticated!
+  end
+
+  def authenticate_user
+    header = request.headers["Authorization"]
+    raise MissingAuthHeader if header.blank?
+    auth_header = header.split(" ").last
+    begin
+      @decoded = JsonWebToken.decode(auth_header)[0]
+      @current_user = User.find(@decoded["user_id"])
+    rescue JWT::DecodeError => e
+      api_error(status: 401, errors: e.message)
+    end
+  end
+
+  def authenticate_user!
+    authenticate_user || UnauthenticatedError
+  end
+
+  def unauthenticated!
+    api_error(status: 401, errors: "not authenticated")
+  end
+
+  def unauthorized!
+    api_error(status: 403, errors: "not authorized")
+  end
+
+  def not_found!
+    api_error(status: 404, errors: "resource not found")
+  end
+
+  def invalid_resource!(errors = [])
+    api_error(status: 422, errors: errors)
+  end
+
+  def paginate(resource)
+    resource.paginate(
+      page: params[:page] || 1, per_page: params[:per_page] || DEFAULT_PER_PAGE
```

Comment:
Can you follow pahination spec from json api spec, this seems slightly different from the specification

Reply:
@tachyons updated, please check!!